src/sabre_swap/mod.rs

```diff
@@ -0,0 +1,147 @@
+// This code is part of Qiskit.
+//
+// (C) Copyright IBM 2022
+//
+// This code is licensed under the Apache License, Version 2.0. You may
+// obtain a copy of this license in the LICENSE.txt file in the root directory
+// of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+//
+// Any modifications or derivative works of this code must retain this
+// copyright notice, and modified files need to carry a notice indicating
+// that they have been altered from the originals.
+
+#![allow(clippy::too_many_arguments)]
+
+pub mod edge_list;
+pub mod qubits_decay;
+pub mod swap_scores;
+
+use ndarray::prelude::*;
+use numpy::IntoPyArray;
+use numpy::PyReadonlyArray2;
+use pyo3::prelude::*;
+use pyo3::wrap_pyfunction;
+use pyo3::Python;
+
+use rayon::prelude::*;
+
+use crate::nlayout::NLayout;
+
+use edge_list::EdgeList;
+use qubits_decay::QubitsDecay;
+use swap_scores::SwapScores;
+
+const EXTENDED_SET_WEIGHT: f64 = 0.5; // Weight of lookahead window compared to front_layer.
+
+#[pyclass]
+pub enum Heuristic {
+    Basic,
+    Lookahead,
+    Decay,
+}
+
+#[pyfunction]
+pub fn sabre_score_heuristic(
+    py: Python,
+    layer: EdgeList,
+    layout: &NLayout,
+    swap_scores: &mut SwapScores,
+    extended_set: EdgeList,
+    distance_matrix: PyReadonlyArray2<f64>,
+    qubits_decay: QubitsDecay,
+    heuristic: &Heuristic,
+) -> PyObject {
+    let dist = distance_matrix.as_array();
+    swap_scores
+        .scores
+        .iter_mut()
+        .for_each(|(swap_qubits, score)| {
+            let mut trial_layout = layout.clone();
+            trial_layout.swap_logical(swap_qubits[0], swap_qubits[1]);
+            *score = score_heuristic(
+                heuristic,
+                &layer.edges,
+                &extended_set.edges,
+                &trial_layout,
+                swap_qubits,
+                &dist,
+                &qubits_decay.decay,
+            );
```

Comment:
I'll give this a try after my current round of benchmarks finish. I think you're right that swapping an extra time should be faster than cloning, especially for larger numbers of qubits because it's basically just two memory swap calls  in `NLayout`. I'm not sure how much of a speed up it'll provide in practice but we'll see.

Reply:
I did this in https://github.com/Qiskit/qiskit-terra/pull/7977/commits/6b5d1cb3dc77694502b4edd818ba6a6d29d8cb6d it honestly didn't make a big difference in the benchmarks mostly because at this point `sabre_score_heuristic()` is very far from being a bottleneck. Looking at the profile for the 50x20 QV `SabreSwap.run()` take ~2.25 sec and `sabre_score_heuristic` is .0667 seconds of that.